src/wasm/wasm-type.cpp

```diff
@@ -961,6 +991,267 @@ std::ostream& operator<<(std::ostream& os, HeapTypeInfo info) {
   WASM_UNREACHABLE("unexpected kind");
 }
 
+struct TypeBuilder::Impl {
+  TypeStore typeStore;
+  HeapTypeStore heapTypeStore;
+
+  struct Entry {
+    bool initialized = false;
+    HeapTypeInfo info = Signature();
+    void set(HeapTypeInfo&& hti) {
+      info = hti;
+      initialized = true;
+    }
+    HeapType get() { return HeapType(uint64_t(&info)); }
+  };
+
+  std::vector<Entry> entries;
+};
+
+TypeBuilder::TypeBuilder(size_t n) {
+  impl = std::make_unique<TypeBuilder::Impl>();
+  impl->entries.resize(n);
+}
+
+TypeBuilder::~TypeBuilder() = default;
+
+void TypeBuilder::setHeapType(size_t i, Signature signature) {
+  assert(i < impl->entries.size() && "Index out of bounds");
+  impl->entries[i].set(signature);
+}
+
+void TypeBuilder::setHeapType(size_t i, const Struct& struct_) {
+  assert(i < impl->entries.size() && "index out of bounds");
+  impl->entries[i].set(struct_);
+}
+
+void TypeBuilder::setHeapType(size_t i, Struct&& struct_) {
+  assert(i < impl->entries.size() && "index out of bounds");
+  impl->entries[i].set(std::move(struct_));
+}
+
+void TypeBuilder::setHeapType(size_t i, Array array) {
+  assert(i < impl->entries.size() && "index out of bounds");
+  impl->entries[i].set(array);
+}
+
+Type TypeBuilder::getTempTupleType(const Tuple& tuple) {
+  return impl->typeStore.canonicalize(tuple);
```

Comment:
How does this work? For a Ref type there is an index i, which gives us the necessary indirection, but here the Tuple contains actual Types..? (There also doesn't seem to be a test for `getTempTupleType`.)

Reply:
This is for when you need to create a tuple of temporary types, for example when constructing a function signature. You can't create a normal tuple type for that because that would allow references to the temporary types to leak out into the global type store. In the worst case, a signature type created later could be canonicalized to a type that refers to the original temporary types which have long since become invalid.

Reply:
I think I see. Is this an internal API then (and therefore not tested as part of the public API)?

Reply:
No, wait, the user needs to create those tuples, for each function..?

Reply:
Right, users need to do this. I forgot to say that I will add a test for this 👍

Reply:
Hmm, why are tuples different than everything else, though? The others all get `size_t i`.

Reply:
Tuples don't have associated HeapTypes, so they don't need to take a heap type index. They still need to be backed by the TypeBuilder's TypeStore, though.